### ISSUE TITLE
Avoid storing large amounts of PushNotification objects

### ISSUE THREAD

**Issue by wolfhechel** [rakanalh/django-pushy#24]
As it is now all PushNotifications are stored and never purged from the database, 
which results in a huge table after a while.

Unless there's a very good reason to store everything I would suggest a change in the way messages are sent:
1. send_push_notification doesn't save a PushNotification object at all, status about sent or not_sent push notifications should be handled in the broker itself.
2. The celerybeat task that sends PushNotifications that hasn't been sent actually removes the PushNotification objects after they're sent.

**Comment by rakanalh** [comment]
Hi @wolfhechel 

I am not sure if i understand you correctly, but PushNotification objects are created and stored in DB in 2 cases:
1) Through admin
2) Through calling [send_push_notification](https://github.com/rakanalh/django-pushy/blob/master/pushy/utils.py#L6).

It seems like you're sending alot of notifications :) which is the case for me as well. However, am not sure if relying on the broker is the better approach. I can think of the models providing some kind of history to the notifications & ability to resend certain notifications by resetting their sent status .. etc. 
However, i think your issue is valid, because what i do is manually go into the DB and delete notifications before certain period of time. 
Would you think that solves your problem?

**Comment by wolfhechel** [comment]
Sure that's one way of doing it, an automatic task could purge them after a certain time.

I can understand that keeping them in the DB would make sure that no notifications are missed out,
so for that reason we could delete objects that are successfully sent and keep the failed ones?

And regarding resending notifications, I'd rather see that as a separate function kept in its own model since it's really hard to find the right Notification in a table with ~900000 rows.

**Comment by rakanalh** [comment]
Yup... i believe the additional value that the DB records provide other than controlling the app flow is informative. So keeping old records is useless in that case.

So if we remove the old push notifications, your table is not going to contain a bunch of unneeded of rows in that case.

In the early version of django-pushy, before it became an app, i extended the model so that it has a 'manual' flag field which indicates whether the push notification was added manually or automatically. The admin interface would filter the list by showing manual stuff only so that all the automated push notifications are hidden for less confusion to whoever is adding content / notifications. This could be added into this current app and in this case we'll have to implement 2 things: a task to purge old notifications and admin interface showing only manually added notifications.

How does that sound?

**Comment by rakanalh** [comment]
Speaking of which, could you please drop me an email, i'd like to start a conversation with you with regards to your setup and some ideas i've been thinking about for django-pushy.

**Comment by wolfhechel** [comment]
Well, something along those lines sounds better yes. 
I think it would be better though to keep that in a separate model though to more clearly separate the functionality.

You could setup a https://gitter.im/ channel for this repository instead and invite me there.

**Comment by rakanalh** [comment]
Ok so here's what i am going to do:

1. Add an config key that let's you enable or disable storing automated push notifications which are inserted into DB using send_push_notification from the utils model.
2. Add a config key to delete DB notifications once they're sent.

In your case, 1 could be False and 2 could be True which means your table is going to be constantly empty unless you're adding notifications from admin.

I see some value in persisting push notifications for users who would like that option.

**Comment by wolfhechel** [comment]
Yeah that should work for all scenarios I can think of!

**Comment by wolfhechel** [comment]
So we haven't followed up on this for the longest time but I'm in need of some change here, as it is right now I need to minimize the I/O on the database, the largest source at this time is pushy.

So my idea here is to keep both variants, the PushNotification model stays as a method to schedule notifications sent from the admin panel, and the send_push_notification() method schedules a notification sent without model interaction.

The drawback with this as you previously mentioned is that we don't have a log anymore, but as I've come to realise after having this code in deployment for a long period is that at least the APNs is inconsistent in it's responses and the sent/not sent flag can't always be trusted.

There's another feature not being utilized in this code for now that is almost impossible with the current model structure, both GCM, FCM and APNS supports batch sends. Although the dispatchers at this point only takes a single token in regard.

What I would like to be able to do is, as an example: `self.request.user.devices.send_notification(payload)`

Basically, manipulate the Device QuerySet as I see fit and schedule notifications sent to all those devices in an optimal way, and an optimal way would be with batch sends to all tokens for the different dispatchers.

**Comment by rakanalh** [comment]
Sorry @wolfhechel, i haven't had the time to actually go about releasing v0.3.0... am also wondering about the status of you py3 branch (i guess you have the same time issue).

I actually thought about this `lf.request.user.devices.send_notification(payload)` very early but doing this would mean that the queryset has to be sent to the celery task at some point and there is no way to persist queryset "queries" that would enable you to reconstruct them later on using Django.

That being said, i guess the flags approach and batch sending should be a priority here. But i can make no promises on when i can get them done... i am thinking in 1 month at minimum because i am currently engaged in some huge project.

**Comment by wolfhechel** [comment]
Well, I had some time to have a look at this. Regarding the "don't save to DB" issue I've mentioned before we altered our dispatcher into using the pushy.tasks.send_single_push_notification task instead, and it works wonders!

About batch sending, this isn't possible at all in APNS2 (the new HTTP/2.0 APNS API), there it's streaming instead. It's still possible in GCM, but that has to be altered again for FCM.

It doesn't have to send the QuerySet, it sends the ID's of the QuerySet resolution instead.

So what I would suggest to do at this point, in order to get django-pushy in shape for 2017 is:
* Assert the Python 3 functionality (I have some work to do here, still)
* Add a Firebase Cloud Messaging dispatcher, Google hasn't fully deprecated GCM yet so all three has to be supported for now.
* Update APNS API to using APNS 2.0 (there's no Python 2 client yet however as far as I can tell).
* Fix batch updates for all dispatchers.

So basically, there's some work to be done but I think it could be managed fairly easy.

Also, I would encourage pushing for a new major release instead, there's bound to be some breaking changes, mainly in how APNS 2.0 works and what keys it expects.
